src/gui/qgsdatasourceselectdialog.cpp

```diff
@@ -0,0 +1,90 @@
+/***************************************************************************
+  qgsdatasourceselectdialog.cpp - QgsDataSourceSelectDialog
+
+ ---------------------
+ begin                : 1.11.2018
+ copyright            : (C) 2018 by Alessandro Pasotti
+ email                : elpaso@itopen.it
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsdatasourceselectdialog.h"
+#include "ui_qgsdatasourceselectdialog.h"
+#include "qgssettings.h"
+#include "qgis.h"
+
+#include <QPushButton>
+
+QgsDataSourceSelectDialog::QgsDataSourceSelectDialog( bool setFilterByLayerType,
+    const QgsMapLayer::LayerType &layerType,
+    QWidget *parent )
+  : QDialog( parent )
+{
+  setupUi( this );
+  setWindowTitle( tr( "Select a data source" ) );
+  QByteArray dlgGeom( QgsSettings().value( QStringLiteral( "/Windows/selectDataSourceDialog/geometry" ), QVariant(), QgsSettings::Section::Gui ).toByteArray() );
+  restoreGeometry( dlgGeom );
+
+  mBrowserModel.initialize();
```

Comment:
I know I'm a little late, but don't we want the app browser instance here?

Reply:
Yes, I though about that and I decided this way because I wanted everything encapsulated in the dialog (I like independent objects that you can just use without too much preparation), the best we can do is to add the browser model as an additional optional argument.

Reply:
Without using app browser model instance we will not have access to Nyall's recent addition to look into project files...

Reply:
ok, I'll deal with that in the bad layers pr, I was waiting for this to be merged and now I can move the the fix layer action (setDataSource and style restore) to app where the model is available.